mountpoint-s3/src/fs.rs

```diff
@@ -55,10 +55,7 @@ enum FileHandleType<Client: ObjectClient, Runtime> {
         request: AsyncMutex<Option<PrefetchGetObject<Client, Runtime>>>,
         etag: ETag,
     },
-    Write {
-        request: AsyncMutex<UploadRequest<Client>>,
-        handle: WriteHandle,
-    },
+    Write(AsyncMutex<UploadRequest<Client, WriteHandle>>),
```

Comment:
I think the `is_in_progress` logic belongs in the file system (it's ultimately about file handle management rather than the "stream"). And moving it here lets us keep using the type system more strongly. Make this `AsyncMutex<Option<UploadRequest<Client, WriteHandle>>>`, and then `complete` can still consume `self` by `take`ing it out of the `Option`. That way the uploader knows it's impossible to call `write` on an already finished upload, so we don't need all the tests for the state management there.

Reply:
The counterpoint is that we do need to handle the state in `fs`. But I gave it a try and it is probably for the best: see #313 (we can then close this draft).